sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -43,11 +52,14 @@ cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
                 + d * j + k)
 
 
-def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
-                              cnp.ndarray[INDEX_T, ndim=1] indices,
-                              cnp.ndarray[INDEX_T, ndim=1] indptr,
-                              INDEX_T d, INDEX_T interaction_only,
-                              INDEX_T degree):
+def _csr_polynomial_expansion(
+    const DATA_T[:] data,
```

Comment:
@jjerphan The code is not compiling if I add const here.

Reply:
This is a bug of Cython. I have opened https://github.com/cython/cython/issues/5230.

I think it is better to use `cython.numeric` (a built-in fused types which combines `integral` and `floating` if I am not mistaken) over `DATA_T`. It looks like it compiles and this should work then.

Reply:
Thank you for investigating and opening up the issue in Cython.
 I'll try to use cython.numeric instead. Thanks for the guidance.

Reply:
Just to clarify
`cython.numeric  # short, int, long, float, double, float complex, double complex`
the numeric type supports more than just integers and floats.

Reply:
Hum, I think the best in the meantime might be to use:
```cython
ctypedef fused DATA_T:
    float
    double
    int
    long
```